torch/export/_trace.py

```diff
@@ -1472,7 +1480,93 @@ def wrapped_fn(*args):
+                        # This has side effect of proxying stuff like
+                        # proxy.node.meta["val"] = extract_val(val) because at that time, torch function
+                        # mode is still active. It seems bad to turn it off inside proxy_tensor.py, so
+                        # I guess we will just rely on DCE for now to remove extra stuff like detach
+                        torch_function_mode_stack = (
+                            torch.overrides._get_current_function_mode_stack()
+                        )
+                        for mode in torch_function_mode_stack:
+                            if isinstance(mode, PreDispatchTorchFunctionMode):
+                                tracer = mode.tracer
+                                proxy = get_proxy_slot(self, tracer).proxy
+                                inner_proxy = tracer.create_proxy(
+                                    "call_function", builtins.getattr, (proxy, attr), {}
+                                )
+                                # Even though the name says tensor tracking, the impl
+                                # suggests it tracks everything?
+                                track_tensor_tree(
+                                    out, inner_proxy, constant=None, tracer=tracer
+                                )
+                return out
+
+            @contextmanager
+            def override_getattribute_for_subclasses(args):
+                """
+                Context manager that temporarily monkey patches
+                tensor.__getattribute__ so that we can intercept it at
+                torch_function layer.
+                """
+
+                # Dictionary that tracks subclass type to original getattr function
+                # and the attributes we can proxy.
+                tensor_type_to_old_getattribute = {}
+                for arg in args:
+                    if is_traceable_wrapper_subclass(arg):
+                        # Query subclass specific attrs
+                        subclass_attrs = set(dir(arg)) - set(dir(torch.Tensor))
+                        attrs_to_proxy = []
+                        for attr in subclass_attrs:
+                            val = getattr(arg, attr)
+                            # If it is a static function or method, we should always inline
+                            if not isinstance(
+                                val, (types.FunctionType, types.MethodType)
+                            ):
+                                attrs_to_proxy.append(attr)
+                        inner_tensors, _ = arg.__tensor_flatten__()
+                        attrs_to_proxy.extend(inner_tensors)
+                        tensor_type_to_old_getattribute[arg.__class__] = (
+                            arg.__class__.__getattribute__,
+                            attrs_to_proxy,
+                        )
+
+                try:
+                    for arg in args:
```

Comment:
Any reason you're looping through `args` a second time here? I think you can just loop through `tensor_type_to_old_getattribute.items()` (the key is the tensor type, which should be equivalent to `arg.__class__`

Reply:
Ah yep true!